[Merged by Bors] - feat(analysis/calculus/times_cont_diff): differentiability of field inverse

urkud opened the pull request:
---
- [x] depends on: #4807

urkud committed:
Snapshot

urkud committed:
Fix compile, reflow some long lines

hrmacbeth commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -1243,6 +1243,11 @@ lemma times_cont_diff_within_at.times_cont_diff_at {n : with_top ℕ}
   times_cont_diff_at 𝕜 n f x :=
 by rwa [times_cont_diff_at, ← times_cont_diff_within_at_inter hx, univ_inter]
 
+lemma times_cont_diff_at.congr_of_eventually_eq {n : with_top ℕ}
+  (h : times_cont_diff_at 𝕜 n f x) (hg : f₁ =ᶠ[𝓝 x] f) :
+  times_cont_diff_at 𝕜 n f₁ x :=
```
Might as well do a `times_cont_within_at` version.

urkud replied:
`times_cont_within_at` version was there. Golfed the proof.

hrmacbeth commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -1628,6 +1633,14 @@ The identity is `C^∞`.
 lemma times_cont_diff_id {n : with_top ℕ} : times_cont_diff 𝕜 n (id : E → E) :=
 is_bounded_linear_map.id.times_cont_diff
 
+lemma times_cont_diff_within_at_id {n : with_top ℕ} {s x} :
+  times_cont_diff_within_at 𝕜 n (id : E → E) s x :=
+times_cont_diff_id.times_cont_diff_within_at
+
+lemma times_cont_diff_at_id {n : with_top ℕ} {x} :
+  times_cont_diff_at 𝕜 n (id : E → E) x :=
+times_cont_diff_id.times_cont_diff_at
```
If adding these, might as well do a `times_cont_diff_on` version`.

urkud replied:
Done.

hrmacbeth commented on src/topology/algebra/module.lean:
```diff
@@ -283,6 +283,7 @@ end
 
 instance : has_one (M →L[R] M) := ⟨id R M⟩
 
+lemma one_def : (1 : M →L[R] M) = id R M := rfl
```
`@[simp]`?  (And the next one?)

urkud replied:
The next one follows from `coe_id'`. I don't think that we should simplify `1` to `id`: what if someone wants to deal with the semiring of endomorphisms as a semiring?

hrmacbeth commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -2380,6 +2403,49 @@ begin
+
+-- TODO: the next few lemmas don't need `𝕜` or `𝕜'` to be complete
+-- A good way to show this is to generalize proofs about `inverse` to a function `f` such that
+-- `∀ᶠ x in 𝓝 a, x * f x = 1`.
```
```suggestion
-- A good way to show this is to generalize `times_cont_diff_at_ring_inverse` to the setting
-- of a function `f` such that `∀ᶠ x in 𝓝 a, x * f x = 1`.
```

hrmacbeth reviewed (commented):
Otherwise LGTM.

urkud committed:
Update src/analysis/calculus/times_cont_diff.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into tcd-lemmas

urkud committed:
refactor(order/filter): make `filter.has_mem semireducible

This way `simp only []` does not simplify `s ∈ f` to `s ∈ f.sets`.

* Add protected simp lemmas `filter.mem_mk` and `filter.mem_sets`.
* Use implicit argument in `filter.mem_generate_iff`.
* Use `∃ t, s ∈ ...` instead of `s ∈ ⋃ t, ...` in
  `filter.mem_infi_finite` and `filter.mem_infi_finite'`.
* Use an implicit argument in `(non/ne_)empty_of_mem_ultrafilter`.

urkud committed:
Merge branch 'filter-mem' into tcd-lemmas

urkud committed:
Merge branch 'tcd-lemmas' of git://github.com/leanprover-community/mathlib into tcd-lemmas

urkud committed:
Merge branch 'master' into tcd-lemmas

urkud committed:
Merge branch 'master' into tcd-lemmas
